File: ColorKalk_Java/src/Classes/ColorType.java
package Classes;

public abstract class ColorType extends Color{
	public ColorType(int red, int green, int blue){
		super(red,green,blue);	
	}
	private ColorType buildColorType(Color c){
		int red=c.getRed(), green=c.getGreen(), blue=c.getBlue();
	    if(c.is_Primary()){
	        return new Primary(red,green,blue);
	    }else if(c.is_Secondary()){
	        return new Secondary(red,green,blue);
	    }else{
	        return new Tertiary(red,green,blue);
	    }
	}
	public ColorType getFirstPrimary(){
		int red=getRed(), green=getGreen(), blue=getBlue();
	    if(red!=0)
	        return new Primary(red,0,0);
	    if(green!=0)
	        return new Primary(0,green,0);
	    else
	        return new Primary(0,0,blue);
	}
	public Color add(Color c2){
		Color c=super.add(c2);
		return buildColorType(c);
	}
	public Color minus(Color c2){
		Color c=super.minus(c2);
		return buildColorType(c);
	}
	public Color complementary(){
		Color c=super.complementary();
		return buildColorType(c);
	}
	public Color MaxIntensity(){
		Color c=super.MaxIntensity();
		return buildColorType(c);
	}
	public Color MinIntensity(){
		Color c=super.MinIntensity();
		return buildColorType(c);
	}
	public Color LucePura(){
		Color c=super.LucePura();
		return buildColorType(c);
	}
	
}

File: ColorKalk_C++/colortype.cpp
#include "colortype.h"
#include "tertiary.h"
#include "primary.h"
#include "secondary.h"

ColorType::ColorType(unsigned int red,unsigned int green,unsigned int blue) : Color(red,green,blue) {}

ColorType::~ColorType() {}

ColorType* ColorType::getFirstPrimary() const{
    unsigned int red=getRed(), green=getGreen(), blue=getBlue();
    if(red!=0)
        return new Primary(red,0,0);
    if(green!=0)
        return new Primary(0,green,0);
    else //rientra in questo caso anche il nero
        return new Primary(0,0,blue);
}

ColorType* ColorType::buildColorType(Color* c) const{
    unsigned int red=c->getRed(), green=c->getGreen(), blue=c->getBlue();
    if(c->is_Primary()){
        return new Primary(red,green,blue);
    }else if(c->is_Secondary()){
        return new Secondary(red,green,blue);
    }else{
        return new Tertiary(red,green,blue);
    }
}

ColorType* ColorType::operator+(const Color& c2) const{
    Color* c=Color::operator+(c2);
    return buildColorType(c);
}

ColorType* ColorType::operator-(const Color& c2) const{
    Color* c=Color::operator-(c2);
    return buildColorType(c);
}

ColorType* ColorType::complementary() const{
    Color* c=Color::complementary();
    return buildColorType(c);
}

ColorType* ColorType::MaxIntensity() const{
    Color* c=Color::MaxIntensity();
    return buildColorType(c);
}

ColorType* ColorType::MinIntensity() const{
    Color* c=Color::MinIntensity();
    return buildColorType(c);
}

ColorType* ColorType::LucePura() const{
    Color* c=Color::LucePura();
    return buildColorType(c);
}

